### DISCUSSION TITLE
Home assistant as backoffice?

### DISCUSSION THREAD

**Discussion by clsgroen** [lbbrhzn/ocpp#283] [Q&A] [answered]
I've been going through the other posts and haven't found my use case yet so hereby a short description and some questions: my neighbour wants a CP with which he can do loadbalancing and with separate KWh counters for his business (lease) EV and a future private EV. There are enough commercial offerings which offer these but they are mostly based on external hardware (and/or license) and a backoffice system with recurring fees so I'm thinking about something homebuilt.

With this OCCP implementation, would it be possible to:
- Implement load balancing using DSMR 5 (if current on a 3x25 phase exceeds 16A then pause or lower the charging rate)?
- Implement green charging using DSMR 5 (set charging rate to the current production)?
- Recognise which RFID/car is charging and make a KWh counter per RFID/car?
- Implement plug&charge based on preconfigured cars and use the above mentioned KWh counter per car?

If not yet possible, would it be possible with extra development or are there restrictions/limits in the protocol (or HA) which prevent this? Any other solutions/ideas/lines of thoughts to make this possible?

**Comment by lbbrhzn** [comment, marked as answer]
These types of applications were the motivation to implement the occp integration.

With ocpp you can control charging rate, but only if your charger supports the smart charging profile. You can use this to implement load balancing and green charging, using an automation. However, you should take into account that most chargers have a minimal charging current, typically 6A / phase.
Some chargers will detect an RFID and set the idTag of the charging request accordingly. Others do not. We currently do not expose the idTag in the ocpp integration, but it would be easy to add. Plug & charge is part of OCPP v2.0. The ocpp integration only supports OCPP v1.6, as most current generation chargers do.

BTW, the Wallbox Pulsar Plus can do load balancing and green charging with a minimum of 460W / phase. It does require an external kWh meter (EM340), and there's a backoffice, but no recurring fee.

**Reply by clsgroen** [reply]
Thanks for your extensive reply. I've advised my neighbour to buy the Wallbox Copper SB so he will be able to use RFID tags for private/office use, using the free myWallbox portal he should be able to report on charged kWh's per RFID/user. 

I know about the EM340 but it's a bit of an ugly/redundant solution if you can connect to your smart meter's P1 and read all the same values there. If you can have HA do the load balancing using OCPP I think that's more elegant (and in the end cheaper): you don't need additional wiring, you save on the purchase of the EM340 and you keep your fuse box tidy.

**Reply by lbbrhzn** [reply]
I agree, but a system with a meter connected directly to your charger will probably be more reliable and quicker to respond for load balancing scenarios than a system controlled by HA.  Also, please note that neighbour may need to add an MID approved meter anyway, I don't think the meter built into the Copper is MID approved. With RS-485 you can connect multiple meters on the same bus. See https://support.wallbox.com/en/knowledge-base/installing-and-configuring-a-mid-power-meter-for-your-charging-infrastructure/

**Reply by clsgroen** [reply]
Might be interesting to create a device (esp or PI) that reads the P1 port and emulates the EM340 output over RS-485. But that's totally out of scope for HA :). 

I'm assuming the reporting from the myWallbox environment (based on the Wallbox internal meters) will suffice since those figures are also the basis for the 'paid backoffice with automatic monthly invoicing' scenario that's in use for most business EV drivers (at least in NL).

**Reply by tipofthesowrd** [reply]
I agree with both points of view. I'm currently in the same situation where I have a Pulsar Plus connected via OCPP to HA to provide load balancing. So far its working fine. It might be less reliable (depending on HA / your server) but I don't think its any way slower if everything is local. 

I'm currently using a ShellyEM to monitor both PV input and the grid. This gives me power reading every few seconds and on each change of the monitored power. This allows me to set the current limit dynamically (for PV only charging) or even switch charging off completely (when cooking for example).

My Pulsar Plus didn't support OCPP until end of december so I already started working on some sort of RS485 implementation on an ESP32. I had it reading out the ShellyEM sensors from HA and then transmit them of over RS485. Not yet in the format of the EM340. The problem for Pulsar Plus is it acts as a RS485 server. Currently ESPhome can only act as server not a client. I implemented the client functions with some custom libraries in ESPhome but nothing stable so far and seeing as OCPP is now supported on my Pulsar Plus I will probably abandon the project.
